cpp/src/arrow/flight/transport_impl.h

```diff
@@ -0,0 +1,220 @@
+/// supporting. The default method implementations return an error.
+class ARROW_FLIGHT_EXPORT ClientTransportImpl {
+ public:
+  virtual ~ClientTransportImpl() = default;
+
+  /// Initialize the client.
+  virtual Status Init(const FlightClientOptions& options, const Location& location,
+                      const arrow::internal::Uri& uri) = 0;
+  /// Close the client. Once this returns, the client is no longer usable.
+  virtual Status Close() = 0;
+
+  virtual Status Authenticate(const FlightCallOptions& options,
+                              std::unique_ptr<ClientAuthHandler> auth_handler);
+  virtual arrow::Result<std::pair<std::string, std::string>> AuthenticateBasicToken(
+      const FlightCallOptions& options, const std::string& username,
+      const std::string& password);
+  virtual Status DoAction(const FlightCallOptions& options, const Action& action,
+                          std::unique_ptr<ResultStream>* results);
+  virtual Status ListActions(const FlightCallOptions& options,
+                             std::vector<ActionType>* actions);
+  virtual Status GetFlightInfo(const FlightCallOptions& options,
+                               const FlightDescriptor& descriptor,
+                               std::unique_ptr<FlightInfo>* info);
+  virtual Status GetSchema(const FlightCallOptions& options,
+                           const FlightDescriptor& descriptor,
+                           std::unique_ptr<SchemaResult>* schema_result);
+  virtual Status ListFlights(const FlightCallOptions& options, const Criteria& criteria,
+                             std::unique_ptr<FlightListing>* listing);
+  virtual Status DoGet(const FlightCallOptions& options, const Ticket& ticket,
+                       std::unique_ptr<ClientDataStream>* stream);
+  virtual Status DoPut(const FlightCallOptions& options,
+                       std::unique_ptr<ClientDataStream>* stream);
+  virtual Status DoExchange(const FlightCallOptions& options,
+                            std::unique_ptr<ClientDataStream>* stream);
+};
+
+/// The implementation of the Flight service, which implements RPC
+/// method handlers that work in terms of the generic interfaces
+/// here.
+///
+/// Transport implementations should implement the necessary
+/// interfaces and call methods of this service.
+class ARROW_FLIGHT_EXPORT FlightServiceImpl {
+ public:
+  explicit FlightServiceImpl(FlightServerBase* base,
+                             std::shared_ptr<MemoryManager> memory_manager)
+      : service_(base), memory_manager_(std::move(memory_manager)) {}
+  Status DoGet(const ServerCallContext& context, const Ticket& request,
+               TransportDataStream* stream);
+  Status DoPut(const ServerCallContext& context, TransportDataStream* stream);
+  Status DoExchange(const ServerCallContext& context, TransportDataStream* stream);
+  FlightServerBase* base() const { return service_; }
+
+ private:
+  FlightServerBase* service_;
```

Comment:
Bad naming here IMHO.

Reply:
Whoops - renamed to base_ since that's the getter's name anyways.